process/block/preprocess/transactions.go

```diff
@@ -1482,53 +1518,60 @@ func (txs *transactions) ProcessMiniBlock(
 		)
 	}()
 
-	numOfOldCrossInterMbs, numOfOldCrossInterTxs := getNumOfCrossInterMbsAndTxs()
+	numOfOldCrossInterMbs, numOfOldCrossInterTxs := postProcessorInfoHandler.GetNumOfCrossInterMbsAndTxs()
 
-	for index := range miniBlockTxs {
+	for txIndex = 0; txIndex < len(miniBlockTxs); txIndex++ {
+		if txIndex <= indexOfLastTxProcessed {
+			continue
+		}
 		if !haveTime() && !haveAdditionalTime() {
-			return processedTxHashes, index, process.ErrTimeIsOut
+			err = process.ErrTimeIsOut
+			break
 		}
 
 		gasProvidedByTxInSelfShard, err = txs.computeGasProvided(
 			miniBlock.SenderShardID,
 			miniBlock.ReceiverShardID,
-			miniBlockTxs[index],
-			miniBlockTxHashes[index],
+			miniBlockTxs[txIndex],
+			miniBlockTxHashes[txIndex],
 			&gasInfo)
 
 		if err != nil {
-			return processedTxHashes, index, err
-		}
-
-		if scheduledMode {
-			txs.gasHandler.SetGasProvidedAsScheduled(gasProvidedByTxInSelfShard, miniBlockTxHashes[index])
-		} else {
-			txs.gasHandler.SetGasProvided(gasProvidedByTxInSelfShard, miniBlockTxHashes[index])
+			break
 		}
 
-		processedTxHashes = append(processedTxHashes, miniBlockTxHashes[index])
-
 		if txs.flagOptimizeGasUsedInCrossMiniBlocks.IsSet() {
 			if gasInfo.totalGasConsumedInSelfShard > maxGasLimitUsedForDestMeTxs {
-				return processedTxHashes, index, process.ErrMaxGasLimitUsedForDestMeTxsIsReached
+				err = process.ErrMaxGasLimitUsedForDestMeTxsIsReached
+				break
 			}
 		}
 
-		txs.saveAccountBalanceForAddress(miniBlockTxs[index].GetRcvAddr())
+		txs.saveAccountBalanceForAddress(miniBlockTxs[txIndex].GetRcvAddr())
 
 		if !scheduledMode {
-			_, err = txs.txProcessor.ProcessTransaction(miniBlockTxs[index])
+			snapshot := txs.handleProcessTransactionInit(postProcessorInfoHandler, miniBlockTxHashes[txIndex])
```

Comment:
can you extract the normal mode processing into a method so we can test this separately?

Reply:
Done